### ISSUE TITLE
About the usability of distances and CRS

### ISSUE THREAD

**Issue by tiagofassoni** [geopandas/geopandas#942]
I'd like to start by thanking you all for your great work. Also, that I'm presenting this issue as a way to better understand the problems underneath so I can help and solve the issues. I'm willing to do the necessary pull requests.

Anyway, I mentor some students of data science and a recurrent problem I saw them (and myself) having is import data with latitude and longitude in geopandas and expecting the method `.distance()` will deliver distances in meters. Most of them end up using geopy, which, in their mind, "does the right thing by default".

So, I'd like to improve the usability in GeoPandas regarding that. Have some suggestions:
- Create a "great circle" function.
- Add in the docs a section about projections and CRS and how to make the distances behave in a known manner.
- Add, in the distance function of the API docs, a link about "this will not necessarily wield a result in meters, please see the FAQ"

Thanks for your attention!

**Comment by jwilson8767** [comment]
For all my projects I end up projecting data to the state plane (units=meters) before using `.distance()` or `.buffer()` I do this so often that I have written my own buffer wrapper function that takes feet, meters, or degrees and attempts to rectify crs issues before applying the buffer to the geometries. I would be happy to contribute this to geopandas if there is interest (perhaps under something like `gpd.safe_buffer(gdf, projected_crs=crs)` method)

Regardless I agree that documentation should be better and **maybe** any usage of `.distance()`, `.buffer()` or even `.area` should show a warning if the crs is not projected.

**Comment by tiagofassoni** [comment]
I think the warnings are a great idea! I think there should be:

- A warning on dataframe creation if a CRS has not been set-up.
- A warning on the distance function saying a CRS has not been set-up.
Also, a helper function for units.
- A helper function to help create dataframes from latitude/longitude.

Btw, what is a state plane? (and this is not a ridiculous question, it
shows I can't just simply work with my dataframe with lat/lon coords and
expect the right thing to work).

**Comment by jwilson8767** [comment]
> - A helper function to help create dataframes from latitude/longitude.
This was just merged in GH-896

Most U.S. States establish (and sometimes codify into law) spatial reference systems for local surveying, which are referred to as "state planes". Usually these are based on the NAD83 datum and are in feet. Some large states (like FL) have several state planes. depending on what part of the state you are working with.

**Comment by andygarfield** [comment]
State plane is a good solution when you are working with a project area that might span a few counties. You could use UTM zones in order to capture a project area that spans multiple states or state plane zones. There is an annoying learning curve to this, I agree.

I'm not sure how a great circle function would fit into the api, though there's probably a way. One problem is that it could only give you distances. If you wanted to get the area of a polygon, you would still need a CRS because, at that point, you would be talking about a 3D surface which conforms to the curvature of the earth.

These are times when I wish the flat-earthers were right.

**Comment by andygarfield** [comment]
State plane is a good solution when you are working with a project area that might span a few counties. You could use UTM zones in order to capture a project area that spans multiple states or state plane zones. There is an annoying learning curve to this, I agree.

I'm not sure how a great circle function would fit into the api, though there's probably a way. One problem is that it could only give you distances. If you wanted to get the area of a polygon, you would still need a CRS because, at that point, you would be talking about a spherical polygon. There is [math for this sort of problem](https://en.wikipedia.org/wiki/Spherical_trigonometry#Area_and_spherical_excess), but it wouldn't be in the rhelm of things shapely or GEOS is concerned with.

These are times when I wish the flat-earthers were right.

**Comment by waeiski** [comment]
I applaud this idea!

I have an idea about taking the idea of mere inclusion of great circles slightly further. A very welcome functionality would be enabling nearest-neighbor type operations that use geodesy/great circles to accurately determine the nearest feature (between points, lines and polygons). For instance, one might have point feature observations across different oceans and determining the nearest coastline/country to the point isn't very easy especially for points in the middle of the ocean and/or near 180 and -180 longitudes. I dream of a day where spatial joins that use nearest-neighbor approach are easily usable within geopandas.

**Comment by jorisvandenbossche** [comment]
Another option we could consider is adding a "geography" type similar to PostGIS (https://postgis.net/docs/using_postgis_dbmanagement.html#PostGIS_Geography).

When you set your geometry column to be "geography" instead of "geometry" (and we can provide easy methods to switch), then the `distance`, `area`, `length` methods would automatically use proper (great circle) algorithms to give correct answers. But at the same time, all other spatial operations would not be available (since those are GEOS based, which only works in the cartesian plane). To have those, you would need to actively convert to an appropriate CRS (or your geodataframe can already have this when reading it from file), similarly as you need to do now to have correct results.

**Comment by jwilson8767** [comment]
@jorisvandenbossche, I like the idea of having an explicit way of knowing whether you can use geodesic functions, but I don't like the idea of renaming the geometry column as some functions (such as sjoin) don't really care what CRS you're working in. Could we maybe use some of the new CRS functionality in pyproj 2 to implement a `gdf.supports_geodesic` attribute/method based on the existing `gdf.crs` attribute and then have a new parameter like `geodesic=True` on the  `distance`, `area`, `length` methods which would then raise an exception when passed a non-geodesic gdf?

**Comment by martinfleis** [comment]
I agree with @jwilson8767 here. Not fan of `geography` column. I was already proposing this kind of solution in #1095, but we are blocked here for a wee while as we haven't implemented pyproj.CRS yet.

**Comment by jorisvandenbossche** [comment]
>  but I don't like the idea of renaming the geometry column

To clarify, it would not "rename" the column, but change the "dtype" (we currently have a "geometry" dtype, and the idea I suggested above was to also have a "geography" dtype). I would keep the default column name at "geometry", although that can of course be somewhat confusing ..

> as some functions (such as sjoin) don't really care what CRS you're working in.

In the idea I mentioned (which you don't necessarily need to like ;)), a function like `sjoin` would not work if you have "geography" dtype data, but only with "geometry" dtype data. Since sjoin needs spatial predicates based on GEOS, those only work in the cartesian plane (wich is what we now assume you have), and thus if you want those to work correctly you need your data in a proper (projected) CRS.

>  have a new parameter like geodesic=True on the distance, area, length methods 

Having such a keyword for those methods is certainly an alternative. The idea of the "geography" dtype would be that you don't need to specify those keywords all the time, if you know that you have lon/lat data and want only "correct" distance calculations. 

Of course, if you want to combine both distance calculations and spatial predicates in a single workflow while using lat/lon data (and in case you don't care about the inaccuracies in the spatial predicates due to using lat/lon data as if it is in the cartesian plane), the such a keyword can be easier than to change dtype.

**Comment by jorisvandenbossche** [comment]
>  a function like sjoin would not work if you have "geography" dtype data, but only with "geometry" dtype data. Since sjoin needs spatial predicates based on GEOS, those only work in the cartesian plane 

Looking at the PostGIS supported functions for "geography" data (https://postgis.net/docs/PostGIS_Special_Functions_Index.html#PostGIS_TypeFunctionMatrix), it seems they do support some predicates (covers, intersects) for "geography". So for those, a spatial join could also work, in principle. But I suppose they have custom algorithms implemented for that (eg it is not available in PROJ/pyproj, they only have distance/length/area)

**Comment by jwilson8767** [comment]
Out of curiosity, beyond the basic geodesic functions, what others would you imagine someday wanting to utilize yourself?

**Comment by Oreilles** [comment]
Is there a way with the current api get meaningful results from the methods .area(), .distance() and .buffer() on a GeoSeries with unprojected CRS, say 4326  ? Reprojecting the dataset is not an option as data is spanning multiple continents.

**Comment by martinfleis** [comment]
@Oreilles There is an ongoing work on a first-class support for spherical geometry that would allow using these methods on unprojected data (see https://github.com/benbovy/spherely) but that is not included in geopandas yet. With the current API, GeoPandas is limited to planar operations.

**Comment by Oreilles** [comment]
This is awesome. Is there any way we can help accelerate the release, either on the Spherely or Geopandas side ?

**Comment by martinfleis** [comment]
If you have a capacity, you can help over in the spherely repo to get it beyond the proof of the concept stage. We need to ensure it reaches at least beta phase before exposing it in geopandas as an experimental feature.

**Comment by VENKAT1pandi** [comment]
Hi, I would like to work on the documentation part of this issue. My proposed scope is to:

- add a short example to the projections user guide, directly after the existing `set_crs()` example, clarifying that `area`, `length`, `distance`, and `buffer` use the active CRS's coordinate units;
- demonstrate assigning a known geographic CRS with `set_crs("EPSG:4326")`, then using `to_crs(gdf.estimate_utm_crs())` before metric calculations for a compact study area; and
- add a short changelog note, if that is appropriate for a documentation-only change.

Would this scope and placement be welcome? I have the implementation prepared, but I will hold off on opening a pull request until there is confirmation and will adjust it based on maintainer feedback.

**Comment by jorisvandenbossche** [comment]
@VENKAT1pandi such a documentation update would be welcome, yes
